sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/throughputControl/controller/container/ThroughputContainerController.java

```diff
@@ -0,0 +1,321 @@
+            .flatMap(response -> {
+                this.targetDatabaseRid = response.getProperties().getResourceId();
+                return Mono.just(this);
+            });
+    }
+
+    private Mono<ThroughputContainerController> resolveContainerResourceId() {
+        return this.targetContainer.read()
+            .flatMap(response -> {
+                this.targetContainerRid = response.getProperties().getResourceId();
+                return Mono.just(this);
+            });
+    }
+
+    private Mono<ThroughputContainerController> resolveContainerMaxThroughput() {
+        return Mono.defer(() -> Mono.just(this.throughputResolveLevel))
+            .flatMap(throughputResolveLevel -> {
+                if (throughputResolveLevel == ThroughputResolveLevel.CONTAINER) {
+                    return this.resolveThroughputByResourceId(this.targetContainerRid)
+                        .onErrorResume(throwable -> {
+                            if (this.isOfferNotConfiguredException(throwable)) {
+                                this.throughputResolveLevel = ThroughputResolveLevel.DATABASE;
+                            }
+
+                            return Mono.error(throwable);
+                        });
+                } else if (throughputResolveLevel == ThroughputResolveLevel.DATABASE) {
+                    return this.resolveThroughputByResourceId(this.targetDatabaseRid)
+                        .onErrorResume(throwable -> {
+                            if (this.isOfferNotConfiguredException(throwable)) {
+                                this.throughputResolveLevel = ThroughputResolveLevel.CONTAINER;
+                            }
+
+                            return Mono.error(throwable);
+                        });
+                }
+
+                // All the underlying throughput control groups are using target throughput,
+                // which is constant value, hence no need to resolve throughput
+                return Mono.empty();
+            })
+            .flatMap(throughputResponse -> {
+                this.updateMaxContainerThroughput(throughputResponse);
+                return Mono.empty();
+            })
+            .retryWhen(
+                // Throughput can be configured on database level or container level
+                // Retry at most 1 time so we can try on database and container both
+                RetrySpec.max(1).filter(throwable -> this.isOfferNotConfiguredException(throwable))
+            ).thenReturn(this);
+    }
+
+    private Mono<ThroughputResponse> resolveThroughputByResourceId(String resourceId) {
+        // TODO: figure out how this work for serveless account
+        // TODO: work item: https://github.com/Azure/azure-sdk-for-java/issues/18776
+        checkArgument(StringUtils.isNotEmpty(resourceId), "ResourceId can not be null or empty");
+        return this.client.queryOffers(resourceId, new CosmosQueryRequestOptions())
+            .single()
```

Comment:
`queryOffers()` returns a `Flux<FeedResponse>`, so calling `single()` on this may fail if there are more than one value in the underlying `Flux`. Make sure either you are only expecting a single value, or if not, just use `flatMap` directly.

Reply:
yup, will only expect one value here, and we are using same in the internal implementation of readThroughput